README
jabberpy  0.3 README Matthew Allum <breakfast@10.am>

Introduction
-------------
jabber.py is a Python module for the jabber instant messaging
protocol. jabber.py deals with the xml parsing and socket code,
leaving the programmer to concentrate on developing quality jabber
based applications with Python.

jabber.py requires at least python 2.0 and the XML expat parser
module ( included in the standard Python distrubution ).

It is developed on Linux but should run happily on other Unix's and
win32.

The library currently supports enough of the protocol to write simple 
jabber clients.

jabber.py is released under the GPL license.

For more infomation see http://jabberpy.sourceforge.net

Whats Needed
------------- 
You need python >=2.0 with the expat modules installed. You can check
for the expat modules by opening up a Python shell ( make sure it 
reports at least version 2 ) and enter 'import xml.parsers.expat' . 
If you get an error you need to install the python expat libs... 

To do this;

For Debian users;

You need the python2-base package installed, then as root;

apt-get install python2-xml

From Python source (Un*x);

Before compiling Python, have a look at the file Modules/Setup in the
distrubution directory. At the bottom of the file you'll find some 
instructions for installing the expat modules, follow them !

For Windows Users;

AFAIK the windows Python Binary ( from python.org ) comes with expat 
support included. If you are using windows see the 'notes' section at
the bottom of this readme.  

Installation
-------------
From the untared distrubution directory; As root run;

python setup.py install

Alternatively just copy xmlstream.py and jabber.py to somewhere in your
PYTHONPATH. 

The examples directory contains some simple jabber based programs which 
use the library. See examples/README

Documentaion
-------------
The modules contain embedded docmentation, use pydoc to present this is 
a nice readable format. 

Also see the protocol documentation on jabber.org and the source code for
the scripts in examples/* . 

Bugs
----
Probably lots :p. If you find any subscribe to the dev list ( see below ) or
drop me a mail - breakfast@10.am

Notes
-----
<> The library should run fine on windows. Unfortunatly the test 
   client will not as it 'selects' on a file handle and 
   windows does not support this. If anyone knows a workaround
   for this please then inform me.

<> A Mailing list for the development of the jabberpy project is 
   available at;

      http://lists.sourceforge.net/lists/listinfo/jabberpy-devel

   Please subscribe and help the project. 






ChangeLog
2001-12-04 12:22  mallum

	* ChangeLog, jabber.py, examples/pygtkjab.py: small fixes to X
	functionality. Yet more pygtkjab improvements

2001-11-20 00:15  mallum

	* jabber.py: component improvements

2001-11-19 18:30  mallum

	* jabber.py, examples/pygtkjab.py: improved error handling slighly,
	added a timeout to sendAndRecieves

2001-11-14 00:07  mallum

	* jabber.py, examples/pygtkjab.py, examples/test_client.py:
	improvements to pygtk, bug squashed in test client too

2001-11-10 23:06  mallum

	* CREDITS, jabber.py, examples/mailchecker.py: Added CREDITS and
	submitted mailchecker.py component

2001-11-09 10:37  mallum

	* jabber.py, examples/pygtkjab.py: iq error bug fixed ( thx brian )
	, pygtk client improvements

2001-11-07 14:07  mallum

	* jabber.py, xmlstream.py, examples/offline.xpm,
	examples/online.xpm, examples/pending.xpm, examples/pygtkjab.py:
	Imporved pygtk client, minor bug fixes to core and component stuff

2001-10-28 01:09  mallum

	* jabber.py, examples/pygtkjab.py: pygtk client much improved. bug
	fixes to jabber.py, mainly roster class

2001-10-16 22:22  mallum

	* examples/: pygtkjab.py, pyqtjab.py: added Arrons pyqt client,
	started work on improving the pygtk client

2001-10-04 00:09  mallum

	* jabber.py, xmlstream.py, docs/index.html: More Small Class
	changes

2001-10-04 00:08  mallum

	* examples/pygtkjab.py: pygtk client fixes for new Class structure

2001-09-24 18:56  mallum

	* jabber.py: More work on a Component class

2001-09-22 01:29  mallum

	* examples/test_client.py, ChangeLog, README, jabber.py,
	xmlstream.py, docs/index.html: re-jigged xmlstream.py classes.
	Renamed jabber.connection ( to Client ). Sketched out a
	jabber.Connection class

2001-09-09 20:58  mallum

	* MANIFEST, README, setup.py, xmlstream.py: Improved README and
	setup.py

2001-09-08 18:09  mallum

	* jabber.py, xmlstream.py, examples/jrpc.py, examples/pygtkjab.py,
	examples/test_client.py: More documenting...

2001-08-29 16:52  mallum

	* ChangeLog, Changelog, Jabber.py, XMLStream.py, jabber.py,
	setup.py, xmlstream.py, docs/index.html, examples/jrpc.py,
	examples/pygtkjab.py, examples/test_client.py: more documentation
	added, better diconnection handling, renaming to lower case of
	modules

2001-08-28 16:23  mallum

	* Jabber.py, docs/howto.tex: began adding documentation

2001-08-27 22:36  mallum

	* Jabber.py, examples/README: Added 'x' tag methods to the protocol
	object, added DJ's suggestion that buildreply keeps the message
	type

2001-08-27 00:53  mallum

	* Changelog, Jabber.py, MANIFEST, XMLStream.py, pygtkjab.py,
	examples/README, examples/jrpc.py, examples/pygtkjab.py,
	examples/test_client.py: Added Changelog, removed dependancie for
	clients to include XMLStream

2001-08-27 00:06  mallum

	* Jabber.py, XMLStream.py, test_client.py, examples/jrpc.py,
	examples/test_client.py: Added XMLStreamNodeBuilder, payload
	methods for iq objects, new jrpc.py example, added examples
	directory

2001-08-25 20:03  mallum

	* Jabber.py, XMLStream.py: added some methods to ease the edition
	of XMLRPC packets

2001-08-14 22:41  mallum

	* XMLStream.py, pygtkjab.py: added a disconnected callback, began
	to sort out gtk app

2001-08-11 15:23  mallum

	* README, setup.py, test_client.py: 0.1 release

2001-08-08 22:14  mallum

	* Jabber.py, test_client.py: more roster improvemnts

2001-07-31 23:29  mallum

	* Jabber.py, test_client.py: Roster class added, JID handling
	improved

2001-07-31 17:03  mallum

	* Jabber.py, pygtkjab.py: added begginings of Roster class

2001-07-26 23:39  mallum

	* Jabber.py, pygtkjab.py: Added initial logon dialog - stilllots to
	do

2001-07-25 00:35  mallum

	* Jabber.py, pygtkjab.py: added dialogs to pygtk client

2001-07-23 23:27  mallum

	* pygtkjab.py: more pygtk work

2001-07-23 21:22  mallum

	* pygtkjab.py: pygtk client reqrite

2001-07-20 16:38  mallum

	* Jabber.py, pygtkjab.py: nothing specfial

2001-07-16 18:49  mallum

	* Jabber.py, pygtkjab.py: more gui client improvements

2001-07-13 18:26  mallum

	* pygtkjab.py, test_client.py: added begginnings of ipaq pygtk
	client

2001-07-13 00:04  mallum

	* Jabber.py, XMLStream.py, test_client.py: JID class created

2001-07-11 17:04  mallum

	* Jabber.py, XMLStream.py, test_client.py: test_client imporved

2001-07-09 17:02  mallum

	* Jabber.py: auth bug fixed, cleaning up

2001-07-06 18:01  mallum

	* Jabber.py, XMLStream.py, test_client.py: improved loggin,

2001-07-05 20:01  mallum

	* Jabber.py, XMLStream.py, test_client.py: 
	more tidying up

2001-07-04 19:33  uid30517

	* Jabber.py, XMLStream.py: added sha auth stuff - thanks DJ

2001-07-04 11:35  uid30517

	* Jabber.py, XMLStream.py: more XMLnodeyness

2001-07-04 00:23  mallum

	* Jabber.py, XMLStream.py: improved object structure - niow all
	protocol types use nodes

2001-07-03 23:01  mallum

	* Jabber.py, XMLStream.py, test_client.py: made more use of nodes
	;)

2001-06-29 18:07  mallum

	* Jabber.py, XMLStream.py, test_client.py: fixed a couple of small
	bugs, started to add some exceptions in ( play around with
	basically )

2001-06-27 18:50  mallum

	* XMLStream.py, Jabber.py, MANIFEST, README, setup.py,
	test_client.py: Initial revision

2001-06-27 18:50  mallum

	* XMLStream.py, Jabber.py, MANIFEST, README, setup.py,
	test_client.py: intial import



jabber/debug.py
# debug.py
#
# Copyright (C) 2003 Jacob Lundqvist
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published
# by the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.

_version_ = '1.4.0'

"""\

Generic debug class

Other modules can always define extra debug flags for local usage, as long as
they make sure they append them to debug_flags

Also its always a good thing to prefix local flags with something, to reduce risk
of coliding flags. Nothing breaks if two flags would be identical, but it might 
activate unintended debugging.

flags can be numeric, but that makes analysing harder, on creation its
not obvious what is activated, and when flag_show is given, output isnt
really meaningfull.

This Debug class can either be initialized and used on app level, or used independantly
by the individual classes.

For samples of usage, see samples subdir in distro source, and selftest
in this code
    
"""


import sys
import time
from string import join

import types


debug_flags = []

color_none = chr(27) + "[0m"
color_black = chr(27) + "[30m"
color_red = chr(27) + "[31m"
color_green = chr(27) + "[32m"
color_brown = chr(27) + "[33m"
color_blue = chr(27) + "[34m"
color_magenta = chr(27) + "[35m"
color_cyan = chr(27) + "[36m"
color_light_gray = chr(27) + "[37m"
color_dark_gray = chr(27) + "[30;1m"
color_bright_red = chr(27) + "[31;1m"
color_bright_green = chr(27) + "[32;1m"
color_yellow = chr(27) + "[33;1m"
color_bright_blue = chr(27) + "[34;1m"
color_purple = chr(27) + "[35;1m"
color_bright_cyan = chr(27) + "[36;1m"
color_white = chr(27) + "[37;1m"


"""
Define your flags in yor modules like this:

from debug import *

DBG_INIT = 'init'                ; debug_flags.append( DBG_INIT )
DBG_CONNECTION = 'connection'    ; debug_flags.append( DBG_CONNECTION )

 The reason for having a double statement wis so we can validate params
 and catch all undefined debug flags
 
 This gives us control over all used flags, and makes it easier to allow
 global debugging in your code, just do something like
 
 foo = Debug( debug_flags )
 
 group flags, that is a flag in it self containing multiple flags should be
 defined without the debug_flags.append() sequence, since the parts are already
 in the list, also they must of course be defined after the flags they depend on ;)
 example:

DBG_MULTI = [ DBG_INIT, DBG_CONNECTION ]



  NoDebug
  -------
  To speed code up, typically for product releases or such
  use this class instead if you globaly want to disable debugging
"""
DBG_INIT = 'init'
debug_flags.append(DBG_INIT)
DBG_ALWAYS = 'always'
debug_flags.append(DBG_ALWAYS)


class NoDebug:

    def __init__(self, *args, **kwargs):
        pass

    def show(self,  *args, **kwargs):
        pass

    def is_active(self, flag):
        pass

    def active_set(self, active_flags=None):
        return 0


LINE_FEED = '\n'


class Debug:

    def __init__(self,
                 #
                 # active_flags are those that will trigger output
                 #
                 active_flags=None,
                 #
                 # Log file should be file object or file namne
                 #
                 log_file=sys.stderr,
                 #
                 # prefix and sufix can either be set globaly or per call.
                 # personally I use this to color code debug statements
                 # with prefix = chr(27) + '[34m'
                 #      sufix = chr(27) + '[37;1m\n'
                 #
                 prefix='DEBUG: ',
                 sufix='\n',
                 #
                 # If you want unix style timestamps,
                 #  0 disables timestamps
                 #  1 before prefix, good when prefix is a string
                 #  2 after prefix, good when prefix is a color
                 #
                 time_stamp=0,
                 #
                 # flag_show should normaly be of, but can be turned on to get a
                 # good view of what flags are actually used for calls,
                 # if it is not None, it should be a string
                 # flags for current call will be displayed
                 # with flag_show as separator
                 # recomended values vould be '-' or ':', but any string goes
                 #
                 flag_show=None,
                 #
                 # If you dont want to validate flags on each call to
                 # show(), set this to 0
                 #
                 validate_flags=1,
                 #
                 # If you dont want the welcome message, set to 0
                 # default is to show welcome if any flags are active
                 welcome=-1,
                 #
                 # Non plain-ascii encodings can benefit from it
                 encoding=None
                 ):

        if type(active_flags) not in [type([]), type(())]:
            print '***'
            print '*** Invalid or oldformat debug param given: %s' % active_flags
            print '*** please correct your param, should be of [] type!'
            print '*** Due to this, full debuging is enabled'
            active_flags = [DBG_ALWAYS]

        if welcome == -1:
            if active_flags and len(active_flags):
                welcome = 1
            else:
                welcome = 0

        self._remove_dupe_flags()
        if log_file:
            if type(log_file) is type(''):
                try:
                    self._fh = open(log_file, 'w')
                except:
                    print 'ERROR: can open %s for writing'
                    sys.exit(0)
            else:  # assume its a stream type object
                self._fh = log_file
        else:
            self._fh = sys.stdout

        if time_stamp not in (0, 1, 2):
            msg2 = '%s' % time_stamp
            raise 'Invalid time_stamp param', msg2
        self.prefix = prefix
        self.sufix = sufix
        self.time_stamp = time_stamp
        self.flag_show = None  # must be initialised after possible welcome
        self.validate_flags = validate_flags
        self.encoding = encoding

        self.active_set(active_flags)
        if welcome:
            self.show('')
            caller = sys._getframe(1)  # used to get name of caller
            try:
                mod_name = ":%s" % caller.f_locals['__name__']
            except:
                mod_name = ""
            self.show('Debug created for %s%s' % (caller.f_code.co_filename,
                                                  mod_name))
            self.show(' flags defined: %s' % join(self.active))

        if type(flag_show) in (type(''), type(None)):
            self.flag_show = flag_show
        else:
            msg2 = '%s' % type(flag_show)
            raise 'Invalid type for flag_show!', msg2

    def show(self, msg, flag=None, prefix=None, sufix=None,
             lf=0):
        """
        flag can be of folowing types:
            None - this msg will always be shown if any debugging is on
            flag - will be shown if flag is active
            (flag1,flag2,,,) - will be shown if any of the given flags 
                               are active

        if prefix / sufix are not given, default ones from init will be used

        lf = -1 means strip linefeed if pressent
        lf = 1 means add linefeed if not pressent
        """

        if self.validate_flags:
            self._validate_flag(flag)

        if not self.is_active(flag):
            return
        if prefix:
            pre = prefix
        else:
            pre = self.prefix
        if sufix:
            suf = sufix
        else:
            suf = self.sufix

        if self.time_stamp == 2:
            output = '%s%s ' % (pre,
                                time.strftime('%b %d %H:%M:%S',
                                              time.localtime(time.time())),
                                )
        elif self.time_stamp == 1:
            output = '%s %s' % (time.strftime('%b %d %H:%M:%S',
                                              time.localtime(time.time())),
                                pre,
                                )
        else:
            output = pre

        if self.flag_show:
            if flag:
                output = '%s%s%s' % (output, flag, self.flag_show)
            else:
                # this call uses the global default,
                # dont print "None", just show the separator
                output = '%s %s' % (output, self.flag_show)

        if type(msg) == type(u'') and self.encoding:
            msg = msg.encode(self.encoding, 'replace')

        output = '%s%s%s' % (output, msg, suf)
        if lf:
            # strip/add lf if needed
            last_char = output[-1]
            if lf == 1 and last_char != LINE_FEED:
                output = output + LINE_FEED
            elif lf == -1 and last_char == LINE_FEED:
                output = output[:-1]
        try:
            self._fh.write(output)
        except:
            # unicode strikes again ;)
            s = u''
            for i in range(len(output)):
                if ord(output[i]) < 128:
                    c = output[i]
                else:
                    c = '?'
                s = s + c
            self._fh.write('%s%s%s' % (pre, s, suf))
        self._fh.flush()

    def is_active(self, flag):
        'If given flag(s) should generate output.'

        # try to abort early to quicken code
        if not self.active:
            return 0
        if not flag or flag in self.active or DBG_ALWAYS in self.active:
            return 1
        else:
            # check for multi flag type:
            if type(flag) in (type(()), type([])):
                for s in flag:
                    if s in self.active:
                        return 1
        return 0

    def active_set(self, active_flags=None):
        "returns 1 if any flags where actually set, otherwise 0."
        r = 0
        ok_flags = []
        if not active_flags:
            # no debuging at all
            self.active = []
        elif type(active_flags) in (types.TupleType, types.ListType):
            flags = self._as_one_list(active_flags)
            for t in flags:
                if t not in debug_flags:
                    print 'Invalid debugflag given', t
                else:
                    ok_flags.append(t)

            self.active = ok_flags
            r = 1
        else:
            # assume comma string
            try:
                flags = active_flags.split(',')
            except:
                self.show('***')
                self.show('*** Invalid debug param given: %s' % active_flags)
                self.show('*** please correct your param!')
                self.show('*** due to this, full debuging is enabled')
                self.active = debug_flags

            for f in flags:
                s = f.strip()
                ok_flags.append(s)
            self.active = ok_flags

        self._remove_dupe_flags()
        return r

    def active_get(self):
        "returns currently active flags."
        return self.active

    def _as_one_list(self, items):
        """ init param might contain nested lists, typically from group flags.

        This code organises lst and remves dupes
        """
        if type(items) <> type([]) and type(items) <> type(()):
            return [items]
        r = []
        for l in items:
            if type(l) == type([]):
                lst2 = self._as_one_list(l)
                for l2 in lst2:
                    self._append_unique_str(r, l2)
            elif l == None:
                continue
            else:
                self._append_unique_str(r, l)
        return r

    def _append_unique_str(self, lst, item):
        """filter out any dupes."""
        if type(item) <> type(''):
            msg2 = '%s' % item
            raise 'Invalid item type (should be string)', msg2
        if item not in lst:
            lst.append(item)
        return lst

    def _validate_flag(self, flags):
        'verify that flag is defined.'
        if flags:
            for f in self._as_one_list(flags):
                if not f in debug_flags:
                    msg2 = '%s' % f
                    raise 'Invalid debugflag given', msg2

    def _remove_dupe_flags(self):
        """
        if multiple instances of Debug is used in same app, 
        some flags might be created multiple time, filter out dupes
        """
        global debug_flags

        unique_flags = []
        for f in debug_flags:
            if f not in unique_flags:
                unique_flags.append(f)
        debug_flags = unique_flags
